City manager fails to properly merge a single city with a single tile that has 2 different city features

### Graphical example
![image](https://github.com/user-attachments/assets/1038542f-dde6-44d0-9c65-3f4fe0e1e1c9)

### Regression test
https://github.com/YetAnotherSpieskowcy/Carcassonne-Engine/commit/d0ce6e39ec5d5ad6d7e489875cbb1465e6b092a8

### Test output
```
$ go test -count=1 ./pkg/game -run TestRegressionGHIssue76  
--- FAIL: TestRegressionGHIssue76 (0.00s)
panic: city manager did not find city: elements.PlacedFeature{Feature:feature.Feature{FeatureType:2, ModifierType:0, Sides:0x30}, Meeple:elements.Meeple{Type:0x0, PlayerID:0x0}} at position position.Position{x:1, y:1} [recovered]
	panic: city manager did not find city: elements.PlacedFeature{Feature:feature.Feature{FeatureType:2, ModifierType:0, Sides:0x30}, Meeple:elements.Meeple{Type:0x0, PlayerID:0x0}} at position position.Position{x:1, y:1}

goroutine 18 [running]:
testing.tRunner.func1.2({0x555d60, 0xc0001843a0})
	/usr/lib/go-1.22/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/lib/go-1.22/src/testing/testing.go:1634 +0x377
panic({0x555d60?, 0xc0001843a0?})
	/usr/lib/go-1.22/src/runtime/panic.go:770 +0x132
github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game/field.(*Field).Expand(0xc000190df8, {0x5c21c8, 0xc0001cc080}, {{0xc000196cf0?, 0x47?, 0x4a?}})
	/home/ubuntu/work/Carcassonne-Engine/pkg/game/field/field.go:109 +0x7c5
github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game.TestRegressionGHIssue76(0xc0001be4e0)
	/home/ubuntu/work/Carcassonne-Engine/pkg/game/field_test.go:236 +0xccf
testing.tRunner(0xc0001be4e0, 0x5921e8)
	/usr/lib/go-1.22/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	/usr/lib/go-1.22/src/testing/testing.go:1742 +0x390
FAIL	github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game	0.003s
FAIL
```

--- a/pkg/game/city/city.go
+++ b/pkg/game/city/city.go
@@ -114,8 +114,16 @@ func (city *City) AddTile(pos position.Position, cityFeatures []elements.PlacedF
 			hasShield = true
 		}
 	}
-	city.features[pos] = cityFeatures
-	city.shields[pos] = hasShield
+	_, tileInCity := city.GetFeaturesFromTile(pos)
+	if !tileInCity {
+		city.features[pos] = cityFeatures
+		city.shields[pos] = hasShield
+	} else {
+		city.features[pos] = append(city.features[pos], cityFeatures...)
+		if !city.shields[pos] {
+			city.shields[pos] = hasShield
+		}
+	}
 	city.checkCompleted()
 }
 

--- a/pkg/game/city/city_manager.go
+++ b/pkg/game/city/city_manager.go
@@ -1,7 +1,6 @@
 package city
 
 import (
-	"fmt"
 	"slices"
 
 	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game/elements"
@@ -139,7 +138,6 @@ func (manager *Manager) UpdateCities(tile elements.PlacedTile) {
 
 	if len(foundCities) > 0 {
 		for _, cityFeature := range tile.GetFeaturesOfType(feature.City) {
-			fmt.Println("aaa")
 			citiesToRemove := make([]int, 0)
 			cityIndexesToJoin := manager.findCitiesToJoin(foundCities, cityFeature.Sides)
 			// join cities

--- a/pkg/game/city/city_manager.go
+++ b/pkg/game/city/city_manager.go
@@ -115,8 +115,9 @@ func (manager *Manager) CanBePlaced(tile elements.PlacedTile, feat elements.Plac
 		// no City features or only has a completely new city
 		return true
 	}
-	citiesToJoin := manager.findCitiesToJoin(foundCities, feat.Sides)
+
 	// this may return an empty list for features that have no cities to join with
+	citiesToJoin := manager.findCitiesToJoin(foundCities, feat.Sides)
 
 	// Check each of the existing cities (if any) found in feature's neighbourhood
 	// We need to check, if they have *any* meeple placed

--- a/pkg/game/city/city_manager.go
+++ b/pkg/game/city/city_manager.go
@@ -1,6 +1,7 @@
 package city
 
 import (
+	"fmt"
 	"slices"
 
 	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game/elements"
@@ -82,32 +83,26 @@ func (manager Manager) findCities(pos position.Position) map[side.Side]int {
 // Checks which cities surrounding positions must be joined after this move.
 // Returns a map of lists of indexes of cities to join. As a key is used a
 // feature that will close the city.
-func (manager Manager) findCitiesToJoin(foundCities map[side.Side]int, tile elements.PlacedTile) map[elements.PlacedFeature][]int {
-	citiesToJoin := map[elements.PlacedFeature][]int{}
-	cityFeatures := tile.GetFeaturesOfType(feature.City)
-	for _, cityFeature := range cityFeatures {
-		var cityIndexesToJoin = []int{}
-		sides := cityFeature.Feature.Sides
-		for _, mask := range side.PrimarySides {
-			if sides.HasSide(mask) {
-				neighbourCityIndex, ok := foundCities[mask]
-				if ok {
-					alreadyChecked := false
-					for _, cityIndex := range cityIndexesToJoin {
-						if cityIndex == neighbourCityIndex {
-							alreadyChecked = true
-							break
-						}
+func (manager Manager) findCitiesToJoin(foundCities map[side.Side]int, sides side.Side) []int {
+	var cityIndexesToJoin = []int{}
+	for _, mask := range side.PrimarySides {
+		if sides.HasSide(mask) {
+			neighbourCityIndex, ok := foundCities[mask]
+			if ok {
+				alreadyChecked := false
+				for _, cityIndex := range cityIndexesToJoin {
+					if cityIndex == neighbourCityIndex {
+						alreadyChecked = true
+						break
 					}
-					if !alreadyChecked {
-						cityIndexesToJoin = append(cityIndexesToJoin, neighbourCityIndex)
-					}
+				}
+				if !alreadyChecked {
+					cityIndexesToJoin = append(cityIndexesToJoin, neighbourCityIndex)
 				}
 			}
 		}
-		citiesToJoin[cityFeature] = cityIndexesToJoin
 	}
-	return citiesToJoin
+	return cityIndexesToJoin
 }
 
 // Checks whether the tile can be placed at given position taking into account
@@ -121,13 +116,12 @@ func (manager *Manager) CanBePlaced(tile elements.PlacedTile, feat elements.Plac
 		// no City features or only has a completely new city
 		return true
 	}
-	citiesToJoin := manager.findCitiesToJoin(foundCities, tile)
+	citiesToJoin := manager.findCitiesToJoin(foundCities, feat.Sides)
 	// this may return an empty list for features that have no cities to join with
-	cToJoin := citiesToJoin[feat]
 
 	// Check each of the existing cities (if any) found in feature's neighbourhood
 	// We need to check, if they have *any* meeple placed
-	for _, cityIndex := range cToJoin {
+	for _, cityIndex := range citiesToJoin {
 		// score report function checks the whole city for placed meeples
 		// and reports any meeples that would be returned which we can use here
 		scoreReport := manager.cities[cityIndex].GetScoreReport()
@@ -144,12 +138,13 @@ func (manager *Manager) UpdateCities(tile elements.PlacedTile) {
 	foundCities := manager.findCities(tile.Position)
 
 	if len(foundCities) > 0 {
-		citiesToRemove := make([]int, 0)
-		citiesToJoin := manager.findCitiesToJoin(foundCities, tile)
-		// join cities
-		for f, cityIndexesToJoin := range citiesToJoin {
+		for _, cityFeature := range tile.GetFeaturesOfType(feature.City) {
+			fmt.Println("aaa")
+			citiesToRemove := make([]int, 0)
+			cityIndexesToJoin := manager.findCitiesToJoin(foundCities, cityFeature.Sides)
+			// join cities
 			if len(cityIndexesToJoin) == 0 {
-				toAppend := []elements.PlacedFeature{f}
+				toAppend := []elements.PlacedFeature{cityFeature}
 				manager.cities = append(manager.cities, NewCity(tile.Position, toAppend))
 			} else {
 				if len(cityIndexesToJoin) > 1 {
@@ -158,19 +153,20 @@ func (manager *Manager) UpdateCities(tile elements.PlacedTile) {
 						citiesToRemove = append(citiesToRemove, cityIndex)
 					}
 				}
-				toAdd := []elements.PlacedFeature{f}
+				toAdd := []elements.PlacedFeature{cityFeature}
 				manager.cities[cityIndexesToJoin[0]].AddTile(tile.Position, toAdd)
 			}
-		}
-		// remove cities that were merged into another city
-		if len(citiesToRemove) > 0 {
-			newCities := make([]City, 0)
-			for index, city := range manager.cities {
-				if !slices.Contains(citiesToRemove, index) {
-					newCities = append(newCities, city)
+
+			// remove cities that were merged into another city
+			if len(citiesToRemove) > 0 {
+				newCities := make([]City, 0)
+				for index, city := range manager.cities {
+					if !slices.Contains(citiesToRemove, index) {
+						newCities = append(newCities, city)
+					}
 				}
+				manager.cities = newCities
 			}
-			manager.cities = newCities
 		}
 	} else {
 		for _, f := range tile.GetFeaturesOfType(feature.City) {

--- a/pkg/game/city/city_manager_test.go
+++ b/pkg/game/city/city_manager_test.go
@@ -104,6 +104,37 @@ func TestUpdateCitiesWhenOneCityClosedSeconedOpen(t *testing.T) {
 	}
 }
 
+func TestUpdateCityWhenTwoFeaturesAdded(t *testing.T) {
+	manager := NewCityManager()
+
+	a := elements.ToPlacedTile(tiletemplates.ThreeCityEdgesConnected().Rotate(1))
+	a.Position = position.New(1, 0)
+	manager.UpdateCities(a)
+
+	b := elements.ToPlacedTile(tiletemplates.ThreeCityEdgesConnected())
+	b.Position = position.New(2, 0)
+	manager.UpdateCities(b)
+
+	c := elements.ToPlacedTile(tiletemplates.FourCityEdgesConnectedShield())
+	c.Position = position.New(2, 1)
+	manager.UpdateCities(c)
+
+	d := elements.ToPlacedTile(tiletemplates.TwoCityEdgesCornerNotConnected().Rotate(1))
+	d.Position = position.New(1, 1)
+	manager.UpdateCities(d)
+
+	if len(manager.cities) != 1 {
+		t.Fatalf("expected %#v, got %#v instead", 1, len(manager.cities))
+	}
+
+	for _, f := range d.GetFeaturesOfType(feature.City) {
+		city, _ := manager.GetCity(d.Position, f)
+		if city == nil {
+			t.Fatalf("not found city feature at side %#v", f.Sides)
+		}
+	}
+}
+
 func TestJoinCitiesOnAdd(t *testing.T) {
 	a := elements.ToPlacedTile(tiletemplates.SingleCityEdgeNoRoads())
 	a.Position = position.New(1, 1)

